modules/host/sector.go

```diff
@@ -0,0 +1,280 @@
+//
+// 12 bytes can be represented as a filepath using 16 base64 characters. This
+// keeps the filesize small and therefore limits the amount of load placed on
+// the filesystem when trying to manage hundreds of thousands or even tens of
+// millions of sectors in a single folder.
+func (h *Host) sectorID(sectorRootBytes []byte) []byte {
+	saltedRoot := crypto.HashAll(sectorRootBytes, h.sectorSalt)
+	id := make([]byte, 16)
+	base64.RawURLEncoding.Encode(id, saltedRoot[:12])
+	return id
+}
+
+// addSector will add a data sector to the host, correctly selecting the
+// storage folder in which the sector belongs.
+func (h *Host) addSector(sectorRoot crypto.Hash, expiryHeight types.BlockHeight, sectorData []byte) error {
+	// Sanity check - sector should have sectorSize bytes.
+	if uint64(len(sectorData)) != sectorSize {
+		build.Critical("incorrectly sized sector passed to addSector in the host")
+		return errors.New("incorrectly sized sector passed to addSector in the host")
+	}
+
+	// Check that there is a storage folder. Though this problem would also be
+	// caught by checking the amount of remaining storage, it is better to
+	// explicit about what is wrong.
+	if len(h.storageFolders) < 1 {
+		return errNoStorage
+	}
+
+	// Check that there is enough room for the sector in at least one storage
+	// folder.
+	enoughRoom := false
+	for _, sf := range h.storageFolders {
+		if sf.SizeRemaining >= sectorSize {
+			enoughRoom = true
+		}
+	}
+	if !enoughRoom {
+		return errInsufficientStorageForSector
+	}
+
+	// Determine which storage folder is going to receive the new sector.
+	err := h.db.Update(func(tx *bolt.Tx) error {
+		// Check whether the sector is a virtual sector.
+		sectorKey := h.sectorID(sectorRoot[:])
+		bsu := tx.Bucket(bucketSectorUsage)
+		usageBytes := bsu.Get(sectorKey)
+		var usage sectorUsage
+		if usageBytes != nil {
+			// usageBytes != nil indicates that this sector is already a in the
+			// database, meaning it's a virtual sector. Add the expiration
+			// height to the list of expirations, and then return.
+			err := json.Unmarshal(usageBytes, &usage)
+			if err != nil {
+				return err
+			}
+			// TODO: check the limit on max number of virtual sector blowup
+			// before doing any more matings - maxVitualSecotrBlowup or
+			// something
+			usage.Expiry = append(usage.Expiry, expiryHeight)
+			usageBytes, err = json.Marshal(usage)
+			if err != nil {
+				return err
+			}
+			return bsu.Put(h.sectorID(sectorRoot[:]), usageBytes)
+		}
+
+		// Try adding the sector to disk. In the event of a failure, the host
+		// will try the next storage folder until there is either a success or
+		// until all options have been exhausted.
+		potentialFolders := h.storageFolders
+		emptiestFolder, emptiestIndex := emptiestStorageFolder(potentialFolders)
+		for emptiestFolder != nil {
+			sectorPath := filepath.Join(h.persistDir, emptiestFolder.uidString(), string(sectorKey))
+			err := ioutil.WriteFile(sectorPath, sectorData, 0700)
```

Comment:
`WriteFile` does not delete the file if an error was encountered, which I assume you want to do